tests/usb_signalling.py

```diff
@@ -5,6 +5,162 @@
 from usb_phy import USB_LINESTATE, USB_TIMINGS
 
 
+class UsbDeviceAttach(UsbEvent):
+    def __init__(self, interEventDelay=0):
+        super().__init__(interEventDelay=interEventDelay)
+
+    def __str__(self):
+        return "DeviceAttach"
+
+    def expected_output(self, bus_speed, offset=0):
+        # return "DeviceAttach\n"
+        return (
+            self.__str__()
+            + "\n"
+            + "DUT entered FS\n"
+            + "Received upstream chirp\n"
+            + "DUT entered HS mode\n"
+        )
+
+    @property
+    def event_count(self):
+        return 1
+
+    def drive(self, usb_phy, bus_speed):
+
+        wait = usb_phy.wait
+        wait_until_ns = usb_phy.wait_until
+        xsi = usb_phy.xsi
+        time = xsi.get_time
+
+        print("DeviceAttach")
+        tConnect_ns = time()
+
+        # Check XcvrSel & TermSel low
+        xcvrsel = xsi.sample_periph_pin(usb_phy._xcvrsel)
+        termsel = xsi.sample_periph_pin(usb_phy._termsel)
+
+        # TODO Drive VBUS and enabled these checks
+        # if xcvrsel == 1:
+        #    print("ERROR: DUT enabled pull up before valid Vbus (XCVRSel)")
+
+        # if termsel == 1:
+        #    print("ERROR: DUT enabled pull up before valid Vbus (TermSel)")
+
+        while True:
+
+            if (time() - tConnect_ns) > USB_TIMINGS["T_SIGATT_US"]:
+                print("ERROR: DUT didnt not assert XcvrSel & TermSel quickly enough")
+
+            # Check device asserts XcvrSel and TermSel before T_SIGATT
+            xcvrsel = xsi.sample_periph_pin(usb_phy._xcvrsel)
+            termsel = xsi.sample_periph_pin(usb_phy._termsel)
+
+            if xcvrsel == 1 and termsel == 1:
+                break
+
+        print("DUT entered FS")
```

Comment:
Is this correct? We pass in bus speed as a parameter at the top, why is it unconditionally 'FS' here? (Unless this isn't bus speed, in which case apologies ;-))

Reply:
All signalling is done at FS, so a HS device must switch back to FS when suspended etc.